dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/AbstractServerCallListener.java

```diff
@@ -68,6 +68,10 @@ public void invoke() {
                     responseObserver.onError(t);
                     return;
                 }
+                if (r.hasException()) {
+                    doOnResponseHasException(r.getException());
+                    return;
+                }
```

Comment:
in this case, why `response.hasException` is false？

Reply:
因为直接invoke的response是同步执行的结果，whenCompleteWithContext里面回调的response是异步的AppResponse，只要provider的实现用了异步，在实际的异步结果里面有异常，就会是外面Invoker的response没异常，whenCompleteWithContext回调的appResponse有异常

Because the response of direct invoke is the result of synchronous execution, the response of the callback in whenCompleteWithContext is an asynchronous AppResponse. As long as the provider implementation uses asynchronous, if there is an exception in the actual asynchronous result, it will be because the response of the external Invoker is not abnormal, and the callback of whenCompleteWithContext is appResponse has exception

Reply:
嗯嗯，那这样看来使用`r.hasException()`替换掉原来的`response.hasException()`应该会更好，因为r才是最终真实的结果

Reply:
OK，我改一下

Reply:
这个地方直接调用 response.hasException() 是不是会导致异步的效果失效呢